angularSTEP/src/app/upload-recipe/upload-recipe.component.ts

```diff
@@ -0,0 +1,106 @@
+import { formattedError } from '@angular/compiler';
+
+@Component({
+  selector: 'app-upload-recipe',
+  templateUrl: './upload-recipe.component.html',
+  styleUrls: ['./upload-recipe.component.scss'],
+})
+
+export class UploadRecipeComponent {
+  constructor(private db: AngularFirestore,
+              private router: Router,) {
+  }
+
+  ingredients: string[] = [];
+  instructions: string[] = [];
+  tools: string[] = [];
+
+  recipeForm = new FormGroup({
+    name: new FormControl('', [Validators.required]),
+    description: new FormControl('', [Validators.required]),
+    tools: new FormControl(this.tools, [Validators.required]),
+    ingredients: new FormControl(this.ingredients, [Validators.required]),
+    instructions: new FormControl(this.instructions, [Validators.required]),
+    extraInfo: new FormControl(''),
+    difficulty: new FormControl('', [Validators.required]),
+  });
+
+
+  addIngredient(newIngredient: string) {
+    var formattedIngredient: string = this.autoCapitalizeFirst(newIngredient);
+    if (formattedIngredient != '' && formattedIngredient != undefined) {
+      this.ingredients.push(formattedIngredient);
+    }
+  }
+
+  addInstruction(newInstruction: string) {
+    var onlyLettersStartInstruction: string =  this.removeNumsAndSymbolsAtStart(newInstruction);
+    if (onlyLettersStartInstruction != '') {
+      var formattedInstruction: string = this.autoCapitalizeFirst(onlyLettersStartInstruction);
+      if (formattedInstruction != '' && formattedInstruction != undefined) {
+        this.instructions.push(formattedInstruction);
+      } 
+    }
+  } 
+
+  addTool(newTool: string) {
+    var formattedTool: string = this.autoCapitalizeFirst(newTool);
+    if (formattedTool != '' && formattedTool != undefined) {
+      this.tools.push(formattedTool);
+    }
+  }
+
+  autoCapitalizeFirst(fullString: string): string {
+    var trimString: string = fullString.trim();
+    if (trimString != '') {
+      return trimString.charAt(0).toUpperCase() + trimString.substring(1);
+    }
+  }
+
+  autoCapitalizeName(fullName: string): string {
+    var trimAndSplitName: string[] = fullName.trim().split(' ');
+    var formattedName: string = '';
+    for (let i = 0; i  < trimAndSplitName.length; i++) {
+      if (trimAndSplitName[i] != '') {
+        formattedName += this.autoCapitalizeFirst(trimAndSplitName[i]);
+        if (i != (trimAndSplitName.length - 1)) {
+          formattedName += ' ';
+        }
+      }
+    }
+    return formattedName;
+  }
+
+  removeNumsAndSymbolsAtStart(originalString: string): string {
+      var startIndex: number = -1;
+
+      for (let i = 0; i < originalString.length; i++) {
+        var currentValue: number = originalString.charCodeAt(i);
+        if ((65 <= currentValue && currentValue <= 90) || (97 <= currentValue && currentValue <= 122)) {
+          return originalString.substring(i);
+        }
+      }
+      if (startIndex == -1) {
+        return '';
```

Comment:
Does something later on catch/warn about adding empty strings?

Reply:
The function removeNumsAndSymbolsAtStart does return an empty string in some cases, but is only called in addInstruction() which then checks the return value. If the return value is an empty string then none of the other commands in addInstruction() execute and the function ends. 

There is no warning or error message about adding empty strings, but the form will not submit if any of the required fields are populated with an empty string, before or after the formatting is applied to their values.